### ISSUE TITLE
hover text and inlay hints for types should be displayed in valid python syntax

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#4004]
for example:

```py
from typing import reveal_type

def fn(value: object):
    if value:
        reveal_type(value) # ~AlwaysFalsy
```

it's not immediately obvious to the user what `~` means. i personally keep misreading it to mean `AlwaysFalsy`. if it was instead displayed as `Not[AlwaysFalsy]` then it would be much more clear.

this is also necessary to address #1625, because insertable inlay hints should be syntactically valid.

**Comment by MichaReiser** [comment]
I don't think we should change this. Similar to the use of `|` for intersections, this is syntax that we'd like to standardize instead.

> this is also necessary to address https://github.com/astral-sh/ty/issues/1625, because insertable inlay hints should be syntactically valid.

I think this is a case where an Inlay hint needs to widen the type (or not show an inlay insertion suggestion). Because inserting `Not[AlwaysFalsy]` is still not particularly useful.

**Comment by DetachHead** [comment]
> I don't think we should change this. Similar to the use of `|` for intersections, this is syntax that we'd like to standardize instead.

i assume you meant `&` not `|`?

personally i find `~` to be less intuitive than `|` and `&`, but that might just be because those are already established in other languages like typescript so i'm used to them. imo `!` makes more sense

in general i believe the more operators the higher the mental burden / learning curve

**Comment by MichaReiser** [comment]
> i assume you meant & not |?

Yes :)

**Comment by AlexWaygood** [comment]
> insertable inlay hints should be syntactically valid

`~AlwaysFalsy` _is_ valid Python syntax

```pycon
>>> ~3
-4
>>> AlwaysFalsy = 3
>>> ~AlwaysFalsy
-4
```

or did you mean to say "valid _type expression_", @DetachHead? That's quite different.

I'm also not sure what your reference to "insertable" means here. Unlike other inlay hint suggestions, the inlay hint in this playground snippet is not insertable by double-clicking on it, because ty knows it will not be valid at runtime: https://play.ty.dev/cfa33935-b136-4041-b76a-043cfde13d33

**Comment by DetachHead** [comment]
> or did you mean to say "valid _type expression_", [@DetachHead](https://github.com/DetachHead)? That's quite different.

honestly i didn't realize that was a valid operator lol. but yeah i do mean type expression

> I'm also not sure what your reference to "insertable" means here. Unlike other inlay hint suggestions, the inlay hint in this playground snippet is not insertable by double-clicking on it, because ty knows it will not be valid at runtime: https://play.ty.dev/cfa33935-b136-4041-b76a-043cfde13d33

oh i didn't realize ty now has insertable inlay hints. still, i think valid type expressions should be preferred when rendering inlay hints because that makes more of them insertable.

for comparison, [pyright's `typePrinter` has a `PythonSyntax` flag](https://github.com/DetachHead/basedpyright/blob/ca31c93749e1ab2546358d6461d76ee2cc9cb669/packages/pyright-internal/src/analyzer/typePrinter.ts#L86) (which seems to also use that term interchangeably with "valid type expression"). in basedpyright i always pass this flag when rendering inlay hints (pylance seems to do the same).

**Comment by AlexWaygood** [comment]
You've mentioned quite a few different things in this issue, but I don't think there's anything actionable for us here:
- Your issue title mentions "valid Python syntax", but `~AlwaysFalsy` _is_ valid Python syntax, so there's no issue here when it comes to syntax highlighting things on hover or in diagnostic messages.
- You mentioned that you'd prefer `Not[AlwaysFalsy]` for our type display rather than `~AlwaysFalsy`. But this would also not be a valid type expression according to the standardised grammar at https://typing.python.org/en/latest/spec/annotations.html#type-and-annotation-expressions, since `Not` is not a standardised special form yet. It's true that `Not[]` is more verbose than `~`, but I see that as a drawback rather than an advantage. `Not[]` would also still not be something that you would be able to auto-insert if you double-clicked on it in an inlay hint, because the definition of `Not` is in `ty_extensions`, which does not exist at runtime
- You mentioned later on that you'd prefer `!` to `~`, but unlike `~`, `!AlwaysFalsy` would not be valid Python syntax. We hope to eventually standardise negation types in the Python type system; proposing new syntax (rather than reusing existing syntax) makes it less likely that the proposal would be accepted. Using invalid Python syntax in our type display now also has the disadvantage that it might not be rendered well by IDEs, as has been discussed elsewhere many times.
- > for comparison, [pyright's typePrinter has a PythonSyntax flag](https://github.com/DetachHead/basedpyright/blob/ca31c93749e1ab2546358d6461d76ee2cc9cb669/packages/pyright-internal/src/analyzer/typePrinter.ts#L86) (which seems to also use that term interchangeably with "valid type expression")
  
  Yes, we have a similar flag in our model, and we always set it to `false` if the display would result in something that would lead to an exception at runtime when you double-click on an inlay hint to insert it. If you find an instance where an inlay hint can be inserted and that then leads to a runtime exception, that's a bug that you should report, but that's not the case here.

**Comment by AlexWaygood** [comment]
I created a followup issue for one concrete improvement that I think we could make to the UX in IDEs here: https://github.com/astral-sh/ty/issues/4037
